src/combinatorics/colex.lean

```diff
@@ -0,0 +1,291 @@
+
+## Tags
+colex, colexicographic, binary
+
+## References
+* https://github.com/b-mehta/maths-notes/blob/master/iii/mich/combinatorics.pdf
+
+## Todo
+Show the subset ordering is a sub-relation of the colex ordering.
+-/
+
+variable {α : Type*}
+
+open finset
+
+/--
+We define this type synonym to refer to the colexicographic ordering on finsets
+rather than the natural subset ordering.
+-/
+@[derive inhabited]
+def finset.colex (α) := finset α
+
+/--
+A convenience constructor to turn a `finset α` into a `finset.colex α`, useful in order to
+use the colex ordering rather than the subset ordering.
+-/
+def finset.to_colex {α} (s : finset α) : finset.colex α := s
+
+/--
+`A` is less than `B` in the colex ordering if the largest thing that's not in both sets is in B.
+In other words, max (A ▵ B) ∈ B (if the maximum exists).
+-/
+instance [has_lt α] : has_lt (finset.colex α) :=
+⟨λ (A B : finset α), ∃ (k : α), (∀ {x}, k < x → (x ∈ A ↔ x ∈ B)) ∧ k ∉ A ∧ k ∈ B⟩
+
+/-- We can define (≤) in the obvious way. -/
+instance [has_lt α] : has_le (finset.colex α) :=
+⟨λ A B, A < B ∨ A = B⟩
+
+/-- The colex ordering on finsets. -/
+def colex.lt [has_lt α] (A B : finset α) : Prop := A.to_colex < B.to_colex
+/-- The colex ordering on finsets. -/
+def colex.le [has_lt α] (A B : finset α) : Prop := A.to_colex ≤ B.to_colex
+
+infix ` <ᶜ `:50 := colex.lt
+infix ` ≤ᶜ `:50 := colex.le
+
+lemma colex.lt_def [has_lt α] (A B : finset α) :
+  A <ᶜ B ↔ ∃ k, (∀ {x}, k < x → (x ∈ A ↔ x ∈ B)) ∧ k ∉ A ∧ k ∈ B :=
+iff.rfl
+lemma colex.le_def [has_lt α] (A B : finset α) :
+  A ≤ᶜ B ↔ A <ᶜ B ∨ A = B :=
+iff.rfl
+
+/-- If everything in A is less than k, we can bound the sum of powers. -/
+lemma binary_sum_nat {k : ℕ} {A : finset ℕ} (h₁ : ∀ {x}, x ∈ A → x < k) :
+  A.sum (pow 2) < 2^k :=
+begin
+  apply lt_of_le_of_lt (sum_le_sum_of_subset (λ t, mem_range.2 ∘ h₁)),
+  have z := geom_sum_mul_add 1 k,
+  rw [geom_series, mul_one, one_add_one_eq_two] at z,
+  rw ← z,
+  apply nat.lt_succ_self,
+end
+
+namespace colex
+
+/-- Strictly monotone functions preserve the colex ordering. -/
+lemma hom {β : Type*} [linear_order α] [decidable_eq β] [preorder β]
+  {f : α → β} (h₁ : strict_mono f) (A B : finset α) :
+  A.image f <ᶜ B.image f ↔ A <ᶜ B :=
+begin
+  simp only [colex.lt_def, not_exists, mem_image, exists_prop, not_and],
+  split,
+  { rintro ⟨k, z, q, k', _, rfl⟩,
+    exact ⟨k', λ x hx, by simpa [h₁.injective] using z (h₁ hx), λ t, q _ t rfl, ‹k' ∈ B›⟩ },
+  rintro ⟨k, z, ka, _⟩,
+  refine ⟨f k, λ x hx, _, _, k, ‹k ∈ B›, rfl⟩,
+  { split,
+    any_goals {
+      rintro ⟨x', hx', rfl⟩,
+      refine ⟨x', _, rfl⟩,
+      rwa ← z _ <|> rwa z _,
+      rwa strict_mono.lt_iff_lt h₁ at hx } },
+  { simp only [h₁.injective, function.injective.eq_iff],
+    exact λ x hx, ne_of_mem_of_not_mem hx ka }
+end
+
+/-- A special case of `colex_hom` which is sometimes useful. -/
+lemma colex_hom_fin {n : ℕ} (A B : finset (fin n)) :
+  A.image (λ n, (n : ℕ)) <ᶜ B.image (λ n, (n : ℕ)) ↔ A <ᶜ B :=
```

Comment:
Shouldn't this be called `colex.hom_fin` instead of `colex.colex_hom_fin`?

Also, this would be a good `simp` lemma, no?
```suggestion
lemma hom_fin {n : ℕ} (A B : finset (fin n)) :
  A.image (λ n, (n : ℕ)) <ᶜ B.image (λ n, (n : ℕ)) ↔ A <ᶜ B :=
```